File: leetcode/python/P0725_SplitLinkedList.py
# https://leetcode.com/problems/split-linked-list-in-parts/

from typing import List, Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def __repr__(self):
        return f"ListNode({self.val}, {self.next})"


class Solution:
    def splitListToParts(
        self, head: Optional[ListNode], k: int
    ) -> List[Optional[ListNode]]:
        # we have to split a linked list in k consecutive linked list parts
        # balanced lengths;
        # e.g. initial list: length n --> parts will have lengths
        #  (n // k) and (n // k + 1) for the first n % k parts
        def get_list_length(head: Optional[ListNode]) -> int:
            crt = head
            list_len = 0
            while crt is not None:
                crt = crt.next
                list_len += 1
            return list_len

        list_length = get_list_length(head)

        result = []
        frag_length = list_length // k

        crt = head
        for i in range(k):
            keep_head = crt
            prev = crt
            crt_len = 0
            while crt is not None and crt_len < frag_length + (
                1 if i < list_length % k else 0
            ):
                prev = crt
                crt = crt.next
                crt_len += 1
            if prev is not None:
                prev.next = None
            result.append(keep_head)

        return result


if __name__ == "__main__":
    s = Solution()
    head = ListNode(
        1,
        ListNode(
            2,
            ListNode(3, next=ListNode(4, ListNode(5, ListNode(6)))),
        ),
    )
    k = 3
    print(s.splitListToParts(head, k))

File: leetcode/cpp/P0725_SplitLinkedList.cpp
// https://leetcode.com/problems/split-linked-list-in-parts/

#include <iostream>
#include <vector>

struct ListNode {
  int val;
  ListNode* next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(next) {}
};

class Solution {
 public:
  std::vector<ListNode*> splitListToParts(ListNode* head, int k) {
    ListNode* h_ = head;
    int listLen = 0;
    while (h_ != nullptr) {
      ++listLen;
      h_ = h_->next;
    }

    std::vector<ListNode*> result;

    h_ = head;
    ListNode* prev = nullptr;
    for (int kk = 0; kk < k; ++kk) {
      int segLen = (listLen / k) + (kk < (listLen % k));
      result.push_back(h_);
      for (int jj = 0; jj < segLen; ++jj) {
        if (h_ != nullptr) {
          prev = h_;
          h_ = h_->next;
        }
      }
      if (prev != nullptr) {
        h_ = prev->next;
        prev->next = nullptr;
        prev = nullptr;
      }
    }
    return result;
  }
};

int main(int argc, char** argv) {
  Solution s;
  s.splitListToParts(nullptr, 3);
  s.splitListToParts(new ListNode(1, new ListNode(2, new ListNode(3))), 5);
  return 0;
}